docs/error-handling.md

```diff
@@ -0,0 +1,190 @@
+and a sign of a serious issue, if the console calls the
+"timeline_detail" mgmt API for a timeline that doesn't exist. ERROR
+would be appropriate in that case. But if the console routinely calls
+the API after deleting a timeline, to check if the deletion has
+completed, then it would be totally normal and INFO or DEBUG level
+would be more appropriate. If a message is logged as an ERROR, but it
+in fact happens frequently in production and never requires any
+action, it should probably be demoted to INFO level.
+
+### Warn
+
+Examples:
+- could not remove temporary file "foobar.temp"
+- unrecognized file "foobar" in timeline directory
+
+Warnings are similar to Errors, in that they should not happen
+when the system is operating normally. The difference between Error and
+Warning is that an Error means that the operation failed, whereas Warning
+means that something unexpected happened, but the operation continued anyway.
+For example, if deleting a file fails because the file already didn't exist,
+it should be logged as Warning.
+
+> **Note:** The python regression tests, under `test_regress`, check the
+> pageserver log after each test for any ERROR and WARN lines. If there are
+> any ERRORs or WARNs that have not been explicitly listed in the test as
+> allowed, the test is marked a failed. This is to catch unexpected errors
+> e.g. in background operations, that don't cause immediate misbehaviour in
+> the tested functionality.
+
+### Info
+
+Info level is used to log useful information when the system is
+operating normally. Info level is appropriate e.g. for logging state
+changes, background operations, and network connections.
+
+Examples:
+- "system is shutting down"
+- "tenant was created"
+- "retrying S3 upload"
+
+### Debug & Trace
+
+Debug and Trace level messages are not Debug-level messages are not
+printed to the log in our normal production configuration, but could
+be enabled for a specific server or tenant, to aid debugging.
+(Although we don't actually have that capability as of this writing).
+
+## Context
+
+We use logging "spans" to hold context information about the current
+operation. Almost every operation happens on a particular tenant and
+timeline, so we enter a span with the "tenant_id" and "timeline_id"
+very early when processing an incoming API request, for example. All
+background operations should also run in a span contain at least those
+two fields, and any other parameters or information that might be
+useful when debugging an error that might happen when performing the
+operation.
+
+## Error message style
+
+PostgreSQL has a style guide for writing error messages:
+
+https://www.postgresql.org/docs/current/error-style-guide.html
+
+Follow that guide when writing error messages in the PostgreSQL
+extension. We don't follow it strictly in the pageserver and
+safekeeper, but the advice in the PostgreSQL style guide is generally
+good, and you can't go wrong by following it.
```

Comment:
+1
Microsoft also provides decent error message writing guidelines and best practices. https://learn.microsoft.com/en-us/windows/win32/debug/error-message-guidelines
We might also give some thought to what an error message review process would look like. I have found such reviews to occur upon request, periodically, or before a feature is released. I am not sure what approach makes the most sense for Neon, but I would be interested in any thoughts on this.

Reply:
> Microsoft also provides decent error message writing guidelines and best practices. https://learn.microsoft.com/en-us/windows/win32/debug/error-message-guidelines

Good advice

> We might also give some thought to what an error message review process would look like. I have found such reviews to occur upon request, periodically, or before a feature is released. I am not sure what approach makes the most sense for Neon, but I would be interested in any thoughts on this.

For the storage components, I think any PR that includes error messages is the right place to review the messages too. A separate review would draw more attention to error messages and make them better, for sure, but I don't think it's worth the effort. (For the user-facing web UI and APIs, error messages are much more prominent and more attention might be warranted.)